### ISSUE TITLE
LG-280: MicroProfile Metrics 3.0 Implementation (part of MP 4.0)

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/open-liberty#11460]
- UFO: https://ibm.box.com/s/ouwlbiaxpeffa2ovw5n06qeetss6u3ad
- FTS: https://github.com/OpenLiberty/open-liberty/issues/15034

**Comment by Channyboy** [comment]
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [ ] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [ ] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [ ] Design / UFO Approval (Alasdair Nottingham) or N/A.
- [ ] No Design / No UFO Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A.
- [ ] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [ ] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [ ] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Legal**
3 weeks before Onboarding
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [ ] Implementation complete. (Epic owner / Feature owner)
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [ ] **FAT** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Demo** - (Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [ ] **Globalization** (Sam Wong - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [ ] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed.
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.


## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

## **Other deliverbles**
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - (Laura Cowen) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by Channyboy** [comment]
UFO public link:
https://ibm.box.com/s/eu9wd2xu01d42cjvb7k33pasvd43vi74

**Comment by Channyboy** [comment]
UFO review on July 21, 2020.

Meeting notes:

UFO link not in the issue.

Page 7 - Question - how does the existing REST metric cause a skew? (3rd bullet - “The existing REST metric ,REST.request which tracks count and cumulative elapsed time duration, can have its data skewed by REST requests that end with an unmapped exception”)

Has this gone far enough? Useful to have a count of how many completed with a particular exception?
Action - continue with discussions for future versions of spec - break out different exceptions for REST metrics.

Emily brought up discussion on metrics going into different namespaces. When REST endpoints are annotated in the application - it will be in the application namespace? REST metrics will be in the base namespace? Grafana has access to all.
Adam Bean (sp). Get user feedback on namespaces - debate on whether there is too much segmenting or not enough.

Page 15 - Don surprised that mapped exception in a tag rather than part of name. Revisit offline. Action item.

Page 19 - typo “produces”

P 23 - Emily question to follow up on.

At end Emily brought up dependencies - Will depend on new mpConfig feature. No backward compatibility.

Jared - is there a change to use Small Rye with mpMetrics-3.0? don’t think so. Considering after 3.0

Migration - Alasdair asked about the migration tools - In the past we have not added microprofile to the migration tools - Previously we brought up adding differences be added to MicroProfile LSP. Revisit with Emily in context of the umbrella feature. #11014

**Comment by cthigh** [comment]
The UFO review recording is now available in box: https://ibm.box.com/s/bhel8ue58hncxxk9qqlu1zsf32dw973y

**Comment by Channyboy** [comment]
Beta blog Request: https://github.com/OpenLiberty/open-liberty/issues/13548

**Comment by Channyboy** [comment]
Public Link (non-expiring) - https://ibm.box.com/s/kcwuyzp842lbn53fd3xc3vtn5awu2qox

**Comment by NottyCode** [comment]
@Channyboy the non-expiring UFO link is not public. Can you fix it so I can approve.

**Comment by Channyboy** [comment]
@NottyCode Link has been updated.

**Comment by Channyboy** [comment]
Link has been updated to expire Dec 31, 2030.
A _non-expiring link_ seems to stop after a few days.

**Comment by Channyboy** [comment]
@gscottj 
Similar to previous MP Metric Features. No change or additions to accessibility.

**Comment by gscottj** [comment]
@steven1046  Per the comment by @Channyboy , please add accessibility approval.

**Comment by Channyboy** [comment]
@chirp1 

For ID approval:

Beta Blog: https://github.com/OpenLiberty/open-liberty/issues/15045
OpenLiberty ID items: https://github.com/OpenLiberty/docs/issues/3277
WS-CD ID Items:  https://github.ibm.com/was-liberty/liberty-docs/issues/2152

Work in progress for a migration blog post from MicroProfile Metrics 2.x to MicroProfile Metrics 3.0

**Comment by Channyboy** [comment]
@skasund 
For an STE approval.
The MicroProfile Metric 3.0 feature has yet to be targetted.
But here is an STE slide. Links will need to be updated as blogs and release links are available.
STE slide here : https://ibm.ent.box.com/file/743648057875

**Comment by samwatibm** [comment]
You've requested focal point approvals.   For globalization approvals, 21001/2 translations are not back yet.  Can you clarify whether there are any translations & I can approve now or whether it was sent & returned previously?  Thanks.

**Comment by Channyboy** [comment]
@samwatibm 
I've sent a shipment request (specifying for 21001 in the note) on November 20th, but it has not been returned yet.

**Comment by samwatibm** [comment]
@Channyboy 
OK thanks.   So 21001 translations won't be back until Jan 4 and I won't approve until you verify that your files are actually in the build after that.   So FYI, globalization approvals won't be done until early January.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**Comment by chirp1** [comment]
Worked out with David updates to the doc issue: https://github.com/OpenLiberty/docs/issues/3277 Approving this feature.

**Comment by Channyboy** [comment]
@samwatibm 
Translation files were delivered with PII shipment for `21.0.0.1`:
https://github.com/OpenLiberty/open-liberty/pull/15475

They have been delivered with `21.0.0.1` instead of `21.0.0.2`  due to the original request targeting `21.0.0.1`. At the time, was unsure which release was targeted and the earliest release was specified.

**Comment by donbourne** [comment]
@Channyboy , please add comment with answers for serviceability approval (even if we don't have SVT complete yet it's helpful for seeing if everything else is ready for this approval)

**Comment by Channyboy** [comment]
@donbourne  The answers were edited in to your post.

**Comment by donbourne** [comment]
Thanks @Channyboy , based on the lack of changes to the user experience, I'm ok with the answers -- just awaiting SVT comment for 3b.
